firebase-crashlytics/src/androidTest/java/com/google/firebase/crashlytics/internal/common/FirebaseCrashlyticsReportManagerTest.java

```diff
@@ -0,0 +1,186 @@
+    final String sessionId = "testSessionId";
+    reportManager.onSessionBegin(sessionId);
+    reportManager.onSessionsFinalize();
+
+    verify(reportPersistence).finalizeReports(sessionId);
+  }
+
+  @Test
+  @SuppressWarnings("unchecked")
+  public void onReportSend_successfulReportsAreDeleted() {
+    final String orgId = "testOrgId";
+    final String sessionId1 = "sessionId1";
+    final String sessionId2 = "sessionId2";
+
+    final AppSettingsData appSettings =
+        new AppSettingsData(null, null, null, null, null, orgId, false);
+
+    final List<CrashlyticsReport> finalizedReports = new ArrayList<>();
+    final CrashlyticsReport mockReport1 = mockReport(sessionId1, orgId);
+    final CrashlyticsReport mockReport2 = mockReport(sessionId2, orgId);
+    finalizedReports.add(mockReport1);
+    finalizedReports.add(mockReport2);
+
+    when(reportPersistence.loadFinalizedReports()).thenReturn(finalizedReports);
+
+    final ArgumentCaptor<OnCompleteListener> testCompletionListener1 =
+        ArgumentCaptor.forClass(OnCompleteListener.class);
+    final ArgumentCaptor<OnCompleteListener> testCompletionListener2 =
+        ArgumentCaptor.forClass(OnCompleteListener.class);
+
+    final Task<CrashlyticsReport> mockTask1 =
+        mockSuccessfulTask(mockReport1, testCompletionListener1);
+    final Task<CrashlyticsReport> mockTask2 =
+        mockFailedTask(new Exception("fail"), testCompletionListener2);
+
+    when(reportSender.sendReport(mockReport1)).thenReturn(mockTask1);
+    when(reportSender.sendReport(mockReport2)).thenReturn(mockTask2);
+
+    reportManager.onReportSend(appSettings);
+
+    verify(reportSender).sendReport(mockReport1);
+    verify(reportSender).sendReport(mockReport2);
+
+    testCompletionListener1.getValue().onComplete(mockTask1);
+    testCompletionListener2.getValue().onComplete(mockTask2);
+
+    verify(reportPersistence).deleteFinalizedReport(sessionId1);
+    verify(reportPersistence, never()).deleteFinalizedReport(sessionId2);
+  }
+
+  private CrashlyticsReport mockReport(String sessionId, String orgId) {
+    final CrashlyticsReport mockReport = mock(CrashlyticsReport.class);
+    final CrashlyticsReport.Session mockSession = mock(CrashlyticsReport.Session.class);
+    when(mockSession.getIdentifier()).thenReturn(sessionId);
+    when(mockReport.getSession()).thenReturn(mockSession);
+    when(mockReport.withOrganizationId(orgId)).thenReturn(mockReport);
+    return mockReport;
+  }
+
+  private Task<CrashlyticsReport> mockSuccessfulTask(
+      CrashlyticsReport result, ArgumentCaptor<OnCompleteListener> completionListener) {
+    return mockTask(result, null, completionListener);
+  }
+
+  private Task<CrashlyticsReport> mockFailedTask(
+      Exception exception, ArgumentCaptor<OnCompleteListener> completionListener) {
+    return mockTask(null, exception, completionListener);
+  }
+
+  @SuppressWarnings("unchecked")
+  private Task<CrashlyticsReport> mockTask(
```

Comment:
Because I'm not super familiar with all the ins and outs of the Tasks API. :) Thanks for letting me know about these! Depending on what I do with the completion callback, these may change anyhow.